File: cloudflare-registry/src/renderers.js
export function renderSubscription(format, nodes) {
  const normalized = (format || "shadowrocket").toLowerCase();
  if (["shadowrocket", "links", "plain"].includes(normalized)) {
    return {
      body: renderLinks(nodes),
      contentType: "text/plain; charset=utf-8"
    };
  }
  if (normalized === "base64") {
    return {
      body: base64Encode(renderLinks(nodes)),
      contentType: "text/plain; charset=utf-8"
    };
  }
  if (["clash", "mihomo"].includes(normalized)) {
    return {
      body: renderClash(nodes),
      contentType: "text/yaml; charset=utf-8"
    };
  }
  if (["shadowrocket-conf", "shadowrocket-config", "sr-conf", "srconfig"].includes(normalized)) {
    return {
      body: renderShadowrocketConfig(nodes),
      contentType: "text/plain; charset=utf-8"
    };
  }
  if (["sing-box", "singbox"].includes(normalized)) {
    return {
      body: JSON.stringify(renderSingBox(nodes), null, 2) + "\n",
      contentType: "application/json; charset=utf-8"
    };
  }
  throw new Error(`unsupported format ${format}`);
}

export function renderLinks(nodes) {
  const links = [];
  for (const node of nodes) {
    for (const protocol of node.protocols || []) {
      if (protocol.type === "trojan") {
        const params = new URLSearchParams();
        if (protocol.sni) {
          params.set("peer", protocol.sni);
          params.set("sni", protocol.sni);
        }
        if (protocol.tls) params.set("security", "tls");
        const query = params.toString();
        links.push(
          `trojan://${encodeURIComponent(protocol.password)}@${node.host}:${protocol.port}${query ? `?${query}` : ""}#${encodeURIComponent(`${node.name}-Trojan`)}`
        );
      }
      if (protocol.type === "shadowsocks") {
        const userInfo = base64UrlEncode(`${protocol.method}:${protocol.password}`);
        links.push(`ss://${userInfo}@${node.host}:${protocol.port}#${encodeURIComponent(`${node.name}-SS`)}`);
      }
    }
  }
  return links.join("\n") + "\n";
}

export function renderClash(nodes) {
  const proxyNames = [];
  const lines = [
    "mixed-port: 7890",
    "allow-lan: false",
    "mode: rule",
    "log-level: info",
    "proxies:"
  ];
  for (const node of nodes) {
    for (const protocol of node.protocols || []) {
      if (protocol.type === "trojan") {
        const name = `${node.name}-Trojan`;
        proxyNames.push(name);
        lines.push(`  - name: ${yamlQuote(name)}`);
        lines.push("    type: trojan");
        lines.push(`    server: ${yamlQuote(node.host)}`);
        lines.push(`    port: ${protocol.port}`);
        lines.push(`    password: ${yamlQuote(protocol.password)}`);
        lines.push(`    sni: ${yamlQuote(protocol.sni || node.host)}`);
        lines.push("    skip-cert-verify: false");
        lines.push("    udp: true");
      }
      if (protocol.type === "shadowsocks") {
        const name = `${node.name}-SS`;
        proxyNames.push(name);
        lines.push(`  - name: ${yamlQuote(name)}`);
        lines.push("    type: ss");
        lines.push(`    server: ${yamlQuote(node.host)}`);
        lines.push(`    port: ${protocol.port}`);
        lines.push(`    cipher: ${yamlQuote(protocol.method)}`);
        lines.push(`    password: ${yamlQuote(protocol.password)}`);
        lines.push("    udp: true");
      }
    }
  }
  lines.push("proxy-groups:");
  lines.push("  - name: PROXY");
  lines.push("    type: select");
  lines.push("    proxies:");
  for (const name of proxyNames) lines.push(`      - ${yamlQuote(name)}`);
  lines.push("      - DIRECT");
  lines.push("rule-providers:");
  lines.push("  direct:");
  lines.push("    type: http");
  lines.push("    behavior: domain");
  lines.push("    url: \"https://cdn.jsdelivr.net/gh/Loyalsoldier/clash-rules@release/direct.txt\"");
  lines.push("    path: ./ruleset/direct.yaml");
  lines.push("    interval: 86400");
  lines.push("  proxy:");
  lines.push("    type: http");
  lines.push("    behavior: domain");
  lines.push("    url: \"https://cdn.jsdelivr.net/gh/Loyalsoldier/clash-rules@release/proxy.txt\"");
  lines.push("    path: ./ruleset/proxy.yaml");
  lines.push("    interval: 86400");
  lines.push("  cncidr:");
  lines.push("    type: http");
  lines.push("    behavior: ipcidr");
  lines.push("    url: \"https://cdn.jsdelivr.net/gh/Loyalsoldier/clash-rules@release/cncidr.txt\"");
  lines.push("    path: ./ruleset/cncidr.yaml");
  lines.push("    interval: 86400");
  lines.push("  lancidr:");
  lines.push("    type: http");
  lines.push("    behavior: ipcidr");
  lines.push("    url: \"https://cdn.jsdelivr.net/gh/Loyalsoldier/clash-rules@release/lancidr.txt\"");
  lines.push("    path: ./ruleset/lancidr.yaml");
  lines.push("    interval: 86400");
  lines.push("rules:");
  lines.push("  - RULE-SET,lancidr,DIRECT");
  lines.push("  - RULE-SET,direct,DIRECT");
  lines.push("  - RULE-SET,cncidr,DIRECT");
  lines.push("  - RULE-SET,proxy,PROXY");
  lines.push("  - GEOIP,CN,DIRECT");
  lines.push("  - MATCH,PROXY");
  return lines.join("\n") + "\n";
}

export function renderShadowrocketConfig(nodes) {
  const proxyNames = [];
  const lines = [
    "[General]",
    "bypass-system = true",
    "skip-proxy = 127.0.0.1, 192.168.0.0/16, 10.0.0.0/8, 172.16.0.0/12, localhost, *.local",
    "",
    "[Proxy]"
  ];
  for (const node of nodes) {
    for (const protocol of node.protocols || []) {
      if (protocol.type === "trojan") {
        const name = shadowrocketName(`${node.name}-Trojan`);
        proxyNames.push(name);
        lines.push(
          `${name} = trojan, ${shadowrocketValue(node.host)}, ${protocol.port}, password=${shadowrocketValue(protocol.password)}, over-tls=true, tls-verification=true, sni=${shadowrocketValue(protocol.sni || node.host)}`
        );
      }
      if (protocol.type === "shadowsocks") {
        const name = shadowrocketName(`${node.name}-SS`);
        proxyNames.push(name);
        lines.push(
          `${name} = ss, ${shadowrocketValue(node.host)}, ${protocol.port}, encrypt-method=${shadowrocketValue(protocol.method)}, password=${shadowrocketValue(protocol.password)}`
        );
      }
    }
  }
  lines.push("");
  lines.push("[Proxy Group]");
  lines.push(`PROXY = select${proxyNames.map((name) => `, ${name}`).join("")}, DIRECT`);
  lines.push("");
  lines.push("[Rule]");
  for (const ruleSetUrl of [
    "https://cdn.jsdelivr.net/gh/blackmatrix7/ios_rule_script@master/rule/Shadowrocket/Lan/Lan.list",
    "https://cdn.jsdelivr.net/gh/blackmatrix7/ios_rule_script@master/rule/Shadowrocket/China/China.list"
  ]) {
    lines.push(`RULE-SET,${ruleSetUrl},DIRECT`);
  }
  lines.push("GEOIP,CN,DIRECT");
  lines.push("FINAL,PROXY");
  return lines.join("\n") + "\n";
}

export function renderSingBox(nodes) {
  const outbounds = [{ type: "direct", tag: "direct" }];
  for (const node of nodes) {
    for (const protocol of node.protocols || []) {
      if (protocol.type === "trojan") {
        outbounds.push({
          type: "trojan",
          tag: `${node.name}-Trojan`,
          server: node.host,
          server_port: protocol.port,
          password: protocol.password,
          tls: {
            enabled: Boolean(protocol.tls),
            server_name: protocol.sni || node.host
          }
        });
      }
      if (protocol.type === "shadowsocks") {
        outbounds.push({
          type: "shadowsocks",
          tag: `${node.name}-SS`,
          server: node.host,
          server_port: protocol.port,
          method: protocol.method,
          password: protocol.password
        });
      }
    }
  }
  return {
    log: { level: "info" },
    dns: { servers: [{ tag: "google", address: "8.8.8.8" }] },
    outbounds,
    route: { final: "direct" }
  };
}

function yamlQuote(value) {
  return `"${String(value).replaceAll("\\", "\\\\").replaceAll("\"", "\\\"")}"`;
}

function shadowrocketName(value) {
  return String(value).replaceAll(",", " ").replaceAll("\n", " ").replaceAll("\r", " ").trim();
}

function shadowrocketValue(value) {
  return String(value ?? "").replaceAll(",", "%2C").replaceAll("\n", "").replaceAll("\r", "");
}

function base64Encode(value) {
  if (typeof btoa === "function") return btoa(value);
  return Buffer.from(value, "utf8").toString("base64");
}

function base64UrlEncode(value) {
  return base64Encode(value).replaceAll("+", "-").replaceAll("/", "_").replaceAll("=", "");
}

File: go-registry/renderers.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/url"
	"strings"
)

func renderSubscription(format string, nodes []Node) (string, string, error) {
	switch strings.ToLower(format) {
	case "", "shadowrocket", "links", "plain":
		return renderLinks(nodes), "text/plain; charset=utf-8", nil
	case "base64":
		links := renderLinks(nodes)
		return base64.StdEncoding.EncodeToString([]byte(links)), "text/plain; charset=utf-8", nil
	case "clash", "mihomo":
		return renderClash(nodes), "text/yaml; charset=utf-8", nil
	case "shadowrocket-conf", "shadowrocket-config", "sr-conf", "srconfig":
		return renderShadowrocketConfig(nodes), "text/plain; charset=utf-8", nil
	case "sing-box", "singbox":
		body, err := renderSingBox(nodes)
		return body, "application/json; charset=utf-8", err
	default:
		return "", "", fmt.Errorf("unsupported format %q", format)
	}
}

func renderLinks(nodes []Node) string {
	var links []string
	for _, node := range nodes {
		for _, p := range node.Protocols {
			switch p.Type {
			case "trojan":
				q := url.Values{}
				if p.SNI != "" {
					q.Set("sni", p.SNI)
					q.Set("peer", p.SNI)
				}
				if p.TLS {
					q.Set("security", "tls")
				}
				link := fmt.Sprintf("trojan://%s@%s:%d", url.QueryEscape(p.Password), node.Host, p.Port)
				if encoded := q.Encode(); encoded != "" {
					link += "?" + encoded
				}
				link += "#" + url.QueryEscape(node.Name+"-Trojan")
				links = append(links, link)
			case "shadowsocks":
				userInfo := base64.RawURLEncoding.EncodeToString([]byte(p.Method + ":" + p.Password))
				links = append(links, fmt.Sprintf("ss://%s@%s:%d#%s", userInfo, node.Host, p.Port, url.QueryEscape(node.Name+"-SS")))
			}
		}
	}
	return strings.Join(links, "\n") + "\n"
}

func renderClash(nodes []Node) string {
	var b strings.Builder
	names := make([]string, 0)
	b.WriteString("mixed-port: 7890\n")
	b.WriteString("allow-lan: false\n")
	b.WriteString("mode: rule\n")
	b.WriteString("log-level: info\n")
	b.WriteString("proxies:\n")
	for _, node := range nodes {
		for _, p := range node.Protocols {
			switch p.Type {
			case "trojan":
				name := node.Name + "-Trojan"
				names = append(names, name)
				fmt.Fprintf(&b, "  - name: %s\n", yamlQuote(name))
				fmt.Fprintf(&b, "    type: trojan\n")
				fmt.Fprintf(&b, "    server: %s\n", yamlQuote(node.Host))
				fmt.Fprintf(&b, "    port: %d\n", p.Port)
				fmt.Fprintf(&b, "    password: %s\n", yamlQuote(p.Password))
				fmt.Fprintf(&b, "    sni: %s\n", yamlQuote(firstNonEmpty(p.SNI, node.Host)))
				fmt.Fprintf(&b, "    skip-cert-verify: false\n")
				fmt.Fprintf(&b, "    udp: true\n")
			case "shadowsocks":
				name := node.Name + "-SS"
				names = append(names, name)
				fmt.Fprintf(&b, "  - name: %s\n", yamlQuote(name))
				fmt.Fprintf(&b, "    type: ss\n")
				fmt.Fprintf(&b, "    server: %s\n", yamlQuote(node.Host))
				fmt.Fprintf(&b, "    port: %d\n", p.Port)
				fmt.Fprintf(&b, "    cipher: %s\n", yamlQuote(p.Method))
				fmt.Fprintf(&b, "    password: %s\n", yamlQuote(p.Password))
				fmt.Fprintf(&b, "    udp: true\n")
			}
		}
	}
	b.WriteString("proxy-groups:\n")
	b.WriteString("  - name: PROXY\n")
	b.WriteString("    type: select\n")
	b.WriteString("    proxies:\n")
	for _, name := range names {
		fmt.Fprintf(&b, "      - %s\n", yamlQuote(name))
	}
	b.WriteString("      - DIRECT\n")
	b.WriteString("rule-providers:\n")
	b.WriteString("  direct:\n")
	b.WriteString("    type: http\n")
	b.WriteString("    behavior: domain\n")
	b.WriteString("    url: \"https://cdn.jsdelivr.net/gh/Loyalsoldier/clash-rules@release/direct.txt\"\n")
	b.WriteString("    path: ./ruleset/direct.yaml\n")
	b.WriteString("    interval: 86400\n")
	b.WriteString("  proxy:\n")
	b.WriteString("    type: http\n")
	b.WriteString("    behavior: domain\n")
	b.WriteString("    url: \"https://cdn.jsdelivr.net/gh/Loyalsoldier/clash-rules@release/proxy.txt\"\n")
	b.WriteString("    path: ./ruleset/proxy.yaml\n")
	b.WriteString("    interval: 86400\n")
	b.WriteString("  cncidr:\n")
	b.WriteString("    type: http\n")
	b.WriteString("    behavior: ipcidr\n")
	b.WriteString("    url: \"https://cdn.jsdelivr.net/gh/Loyalsoldier/clash-rules@release/cncidr.txt\"\n")
	b.WriteString("    path: ./ruleset/cncidr.yaml\n")
	b.WriteString("    interval: 86400\n")
	b.WriteString("  lancidr:\n")
	b.WriteString("    type: http\n")
	b.WriteString("    behavior: ipcidr\n")
	b.WriteString("    url: \"https://cdn.jsdelivr.net/gh/Loyalsoldier/clash-rules@release/lancidr.txt\"\n")
	b.WriteString("    path: ./ruleset/lancidr.yaml\n")
	b.WriteString("    interval: 86400\n")
	b.WriteString("rules:\n")
	b.WriteString("  - RULE-SET,lancidr,DIRECT\n")
	b.WriteString("  - RULE-SET,direct,DIRECT\n")
	b.WriteString("  - RULE-SET,cncidr,DIRECT\n")
	b.WriteString("  - RULE-SET,proxy,PROXY\n")
	b.WriteString("  - GEOIP,CN,DIRECT\n")
	b.WriteString("  - MATCH,PROXY\n")
	return b.String()
}

func renderShadowrocketConfig(nodes []Node) string {
	var (
		b     strings.Builder
		names []string
	)
	b.WriteString("[General]\n")
	b.WriteString("bypass-system = true\n")
	b.WriteString("skip-proxy = 127.0.0.1, 192.168.0.0/16, 10.0.0.0/8, 172.16.0.0/12, localhost, *.local\n")
	b.WriteString("\n")
	b.WriteString("[Proxy]\n")
	for _, node := range nodes {
		for _, p := range node.Protocols {
			switch p.Type {
			case "trojan":
				name := shadowrocketName(node.Name + "-Trojan")
				names = append(names, name)
				fmt.Fprintf(&b, "%s = trojan, %s, %d, password=%s, over-tls=true, tls-verification=true, sni=%s\n",
					name,
					shadowrocketValue(node.Host),
					p.Port,
					shadowrocketValue(p.Password),
					shadowrocketValue(firstNonEmpty(p.SNI, node.Host)),
				)
			case "shadowsocks":
				name := shadowrocketName(node.Name + "-SS")
				names = append(names, name)
				fmt.Fprintf(&b, "%s = ss, %s, %d, encrypt-method=%s, password=%s\n",
					name,
					shadowrocketValue(node.Host),
					p.Port,
					shadowrocketValue(p.Method),
					shadowrocketValue(p.Password),
				)
			}
		}
	}
	b.WriteString("\n")
	b.WriteString("[Proxy Group]\n")
	b.WriteString("PROXY = select")
	for _, name := range names {
		fmt.Fprintf(&b, ", %s", name)
	}
	b.WriteString(", DIRECT\n")
	b.WriteString("\n")
	b.WriteString("[Rule]\n")
	for _, ruleSetURL := range []string{
		"https://cdn.jsdelivr.net/gh/blackmatrix7/ios_rule_script@master/rule/Shadowrocket/Lan/Lan.list",
		"https://cdn.jsdelivr.net/gh/blackmatrix7/ios_rule_script@master/rule/Shadowrocket/China/China.list",
	} {
		fmt.Fprintf(&b, "RULE-SET,%s,DIRECT\n", ruleSetURL)
	}
	b.WriteString("GEOIP,CN,DIRECT\n")
	b.WriteString("FINAL,PROXY\n")
	return b.String()
}

func renderSingBox(nodes []Node) (string, error) {
	outbounds := []map[string]any{{"type": "direct", "tag": "direct"}}
	for _, node := range nodes {
		for _, p := range node.Protocols {
			switch p.Type {
			case "trojan":
				outbounds = append(outbounds, map[string]any{
					"type":        "trojan",
					"tag":         node.Name + "-Trojan",
					"server":      node.Host,
					"server_port": p.Port,
					"password":    p.Password,
					"tls": map[string]any{
						"enabled":     p.TLS,
						"server_name": firstNonEmpty(p.SNI, node.Host),
					},
				})
			case "shadowsocks":
				outbounds = append(outbounds, map[string]any{
					"type":        "shadowsocks",
					"tag":         node.Name + "-SS",
					"server":      node.Host,
					"server_port": p.Port,
					"method":      p.Method,
					"password":    p.Password,
				})
			}
		}
	}
	cfg := map[string]any{
		"log": map[string]any{"level": "info"},
		"dns": map[string]any{
			"servers": []map[string]any{{"tag": "google", "address": "8.8.8.8"}},
		},
		"outbounds": outbounds,
		"route": map[string]any{
			"final": "direct",
		},
	}
	b, err := json.MarshalIndent(cfg, "", "  ")
	if err != nil {
		return "", err
	}
	return string(b) + "\n", nil
}
